### DISCUSSION TITLE
Completing the intellisense-enabled scripting experience

### DISCUSSION THREAD

**Discussion by bernatagulloesbrina** [TabularEditor/TabularEditor3#425] [🙋♀️TE3 Q&A] [answered]
Hi, 

Authoring C# scripts in Visual Studio or VSCode referencing TOMWrapper.dll is great as we get complete intellisense
However, there are a few bits that apparently are not part of that DLL 

"Selected" is maybe the most important one. How can we replicate an object with the same name that provides intellisense on the properties and methods of Selected? 

For Model we just created a variable of Model type named "Model" and it works nice, however not sure how to proceed with Selected 

Other hold-outs are the functions Info Error Output 
not an expert c#er but maybe we could copy some function specification somewhere so that at least is not underlined in red when we type in Visual Studio ?

Thank you! I hope it makes sense

**Comment by otykier** [comment, marked as answer]
Short term, I think you can get the IntelliSense you’re asking for by adding a reference to TabularEditor.exe (TE2).

- `Selected` is an object of type [TabularEditor.UI.UITreeSelection](https://github.com/TabularEditor/TabularEditor/blob/885ff9f84f6497b7958a402854d959d747e0e0a2/TabularEditor/UI/UITreeSelection.cs#L264).
- The global methods (`Info`, `Output`, etc.) are defined [here](https://github.com/TabularEditor/TabularEditor/blob/master/TabularEditor/Scripting/ScriptHelper.cs), so adding `using TabularEditor.Scripting;` should at least allow you to use the extension methods. You can’t refer the methods directly (that is, without qualifying them with `ScriptHelper` or without using them as extension methods). This is only possible when the script is executed in TE because TE has a precompiler that transforms the script before sending it to the C# compiler, i.e. transforming `Info("Hello World");` to `ScriptHelper.Info("Hello World");`.

Longer term, we plan to have full IntelliSense-like features in TE3 when writing C# scripts.

**Reply by samaguire** [reply]
Thanks Daniel! I'm looking forward to TE3 IntelliSense. Is there also another reference that will fix `Selected.Columns` as per below screen snip?

![image](https://user-images.githubusercontent.com/62320770/154162818-9dedd6f1-c9a8-4c0f-8b98-cd111f9cf325.png)

**Reply by samaguire** [reply]
Sorry, that was stupid of me 🤦‍♂️. Got it working after re-reading this.

![image](https://user-images.githubusercontent.com/62320770/154173572-b30c08df-3bdb-46a7-b96d-9338dcb24f7d.png)

**Reply by otykier** [reply]
Awesome! Let me know if you require any additional assistance.
